synapse/util/versionstring.py

```diff
@@ -31,11 +35,11 @@ def get_version_string(module) -> str:
         str
     """
 
-    cached_version = getattr(module, "_synapse_version_string_cache", None)
-    if cached_version:
+    cached_version = version_cache.get(module)
+    if cached_version is not None:
         return cached_version
 
-    version_string = module.__version__
+    version_string = getattr(module, "__version__", "<unknown>")
```

Comment:
I'd vote we keep the behaviour where we raise an exception if the version is unknown. Mainly because let's not change behaviour in a PR that adds type hints.

Reply:
That makse sense. I got tunnel vision trying to make mypy pass.

Reply:
Fair enough. Are you happy to keep the cache in a dictionary in this module (as opposed to adding an attribute to each module?)